runtime-sdk/src/modules/accounts/mod.rs

```diff
@@ -602,3 +674,58 @@ impl module::BlockHandler for Module {
         }
     }
 }
+
+impl module::InvariantHandler for Module {
+    /// Check invariants.
+    fn check_invariants<C: Context>(ctx: &mut C) -> Result<(), CoreError> {
+        // All account balances should sum up to the total supply for their
+        // corresponding denominations.
+
+        let balances = Self::get_all_balances(ctx.runtime_state()).unwrap();
+        let total_supplies = Self::get_total_supplies(ctx.runtime_state()).unwrap();
+
+        // First, compute total supplies based on account balances.
+        let mut computed_ts: BTreeMap<token::Denomination, token::Quantity> = BTreeMap::new();
+
+        for bals in balances.values() {
+            for (den, amt) in bals {
+                computed_ts
+                    .entry(den.clone())
+                    .and_modify(|a| *a += amt)
+                    .or_insert_with(|| amt.clone());
+            }
+        }
+
+        // Now check if the computed and given total supplies match.
+        for (den, ts) in &total_supplies {
+            let computed;
+            match computed_ts.remove(&den) {
+                Some(c) => computed = c,
+                // Total supplies have a denomination that we didn't encounter
+                // when computing total supplies based on account balances.
+                None => {
+                    return Err(CoreError::InvariantViolation(
+                        "unexpected denomination".to_string(),
+                    ))
+                }
+            }
```

Comment:
You could do something like (preserve the comment ofc):
```suggestion
            let computed = computed_ts.remove(&den).ok_or(CoreError::InvariantViolation(...))?;
```